hathor/p2p/sync_v2/payloads.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2023 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from pydantic import validator
+
+from hathor.utils.pydantic import BaseModel
+
+
+class PayloadBaseModel(BaseModel):
+
+    @classmethod
+    def convert_hex_to_bytes(cls, value: str | bytes) -> bytes:
+        if isinstance(value, str):
+            return bytes.fromhex(value)
+        elif isinstance(value, bytes):
+            return value
```

Comment:
This behavior is a bit surprising to me. Looking only at the signature I would have assumed the that it supports encoded/unencoded characters, as in if bytes is given it would still try to hex decode, but assume ascii (or utf-8, shouldn't matter for hex chars).

Is there any practical application for not doing anything if bytes are given? Otherwise I'd rather have it more straightforward and only accept str.

Reply:
Yes, there's a practical application. The current implementation is very flexible and supports all the following object creation. Notice that the validator is always executed, no matter how we create the object.

```
> GetNextBlocksPayload(start_hash=b1.hash, end_hash=b2.hash)
> GetNextBlocksPayload(start_hash=b1.hash_hex, end_hash=b2.hash_hex)
> GetNextBlocksPayload.parse_raw(json.dumps({'start_hash': b1.hash_hex, 'end_hash': b2.hash_hex}))
```

Reply:
If we already do this in other cases I think it's fine to keep the pattern, otherwise this flexibility feels error prone to me. Specially if we're not making use of the flexibility.

Reply:
We use it. To send information, I use `GetNextBlocksPayload(start_hash=b1.hash, end_hash=b2.hash)` to generate a payload and, at the receiver, I use `GetNextBlocksPayload.parse_raw(payload)` to parse the payload.

Reply:
Got it. Seems a bit weird to me that `GetNextBlocksPayload(start_hash=b1.hash, end_hash=b2.hash)` would call effectively deserialization methods, I'm not too familiar with pydantic.